src/mbgl/text/cross_tile_symbol_index.hpp

```diff
@@ -58,7 +58,7 @@ class CrossTileSymbolIndex {
 public:
     CrossTileSymbolIndex();
 
-    bool addLayer(RenderSymbolLayer&, float lng);
+    bool addLayer(const RenderLayer&, float lng);
```

Comment:
Yes, that would be the goal. Do you see any issues with that approach?
Eventually, to achieve https://github.com/mapbox/mapbox-gl-native/issues/13276, there would need to be clean set of interfaces between core components (e.g., renderer)  and layers.

Reply:
We treat each layer module as a library exposing only the generic interfaces, so every layer module just exposes the `RenderLayer` interface , thus the inherited `RenderSymbolLayerInterface` would not be visible for the `gl-native` core.
And as far as I see, generic `RenderLayer` API is enough for the client code (CrossTileSymbolIndex and Placement).

@ChrisLoer could you please give more details on how `RenderSymbolLayerInterface` could be a part of the generic API?

Reply:
It doesn't make much difference now because a custom `RenderLayer` that renders symbols is still required to contain a `SymbolBucket`, which tightly couples it to the existing symbol implementation. But to allow really customized symbol layers, we'll need to be able to provide alternate bucket implementations that conform to the interface currently expected from `SymbolBucket` (they have to expose collision information, including collision debug vertices, they need to have a concept of keyed "symbol instances" that can have opacity updates applied to them, etc.). At a certain point, I think we have to codify those relatively complex interface requirements in the exposed module API -- it would be too much of a minefield to just give the `RenderLayer` abstraction with a bunch of documentation for how symbol layers need to behave, and then fail at runtime in confusing ways if a custom symbol layer didn't conform to one of those expectations.

> And as far as I see, generic RenderLayer API is enough for the client code (CrossTileSymbolIndex and Placement).

I agree that up to the point of returning a `SymbolBucket` it's pretty simple to conform so the generic `RenderLayer` interface is probably enough. I think if there were a generic `SymbolBucket` as part of the public API and the rule was "you have to provide an implementation of `SymbolBucket` if you're returning `hasSymbols -> true`", that would be enough.

Reply:
@ChrisLoer  thanks for detailed explanation. `SymbolBucket` has rather large API surface and it is tightly coupled with other classes. It would be quite challenging to make symbol related code generic, in scope of this PR. I can try adding interface as you proposed.

Reply:
@ChrisLoer what do you think about `RenderLayerSymbolInterface`? Is it acceptable to preserve type safety?

Reply:
Yeah, I think it looks good.